fix(aio): only remove the header-link from toc titles

petebacondarwin opened the pull request:
The previous approach just removed the first `a` tag that
was found, but now that the header-link anchor is not at
the start of the heading, it could fail.

Closes #22493

mary-poppins commented:
You can preview 7c69b2e at https://pr22504-7c69b2e.ngbuilds.io/.

gkalpak commented on aio/src/app/shared/toc.service.ts:
```diff
@@ -53,15 +53,15 @@ export class TocService {
-    const a = this.document.createElement('a') as HTMLAnchorElement;
-    a.innerHTML = heading.innerHTML;
-    const anchorLink = a.querySelector('a');
+    const div = this.document.createElement('div') as HTMLAnchorElement;
```
The cast to `HTMLAnchorElement` is wrong (and seems unnecessary anyway).

petebacondarwin commented:
Actually, I think this PR is wrong as it stands as we should really be removing all links from the HTML that is copied over. Otherwise you have links inside a link which doesn't make sense, right?

petebacondarwin committed:
fix(aio): remove all links from toc titles

The previous approach just removed the first `a` tag that
was found, but now that the header-link anchor is not at
the start of the heading, it could fail.

Closes #22493

petebacondarwin commented on aio/src/app/shared/toc.service.spec.ts:
```diff
@@ -352,13 +349,20 @@ class TestDomSanitizer {
+      throw new Error('$lastInfo is not yet defined. You must call `spyOn` first.');
+    }
+    return this.$$lastInfo;
+  }
```
All this stuff is just to get around strict null checks, which are not yet turned on it seems.

petebacondarwin commented:
@gkalpak PTAL

mary-poppins commented:
You can preview dd39032 at https://pr22504-dd39032.ngbuilds.io/.

gkalpak commented on aio/src/app/shared/toc.service.ts:
```diff
@@ -53,15 +53,25 @@ export class TocService {
-      a.removeChild(anchorLink);
+    const div: HTMLDivElement = this.document.createElement('div');
+    div.innerHTML = heading.innerHTML;
+    const anchorLinks: NodeListOf<HTMLAnchorElement> = div.querySelectorAll('a');
```
OOC: Are the `HTMLDivElement` (above) and `NodeListOf<HTMLAnchorElement>` types necessary? Latest TS can infer them. (Maybe time to upgrade :grin:)

petebacondarwin replied:
'tis true but we haven't upgraded yet!

petebacondarwin commented:
Just realised that I need to change the commit message - since the solution is now different. I will rebase and squash when doing so.

mary-poppins commented:
You can preview 69fd953 at https://pr22504-69fd953.ngbuilds.io/.

petebacondarwin commented:
I accidentally deleted the branch for this PR, which closed this PR. I reinstated the branch but I don't seem to be able to reopen the PR. I will open a new one.
